python/ray/remote_function.py

```diff
@@ -254,6 +264,18 @@ def _remote(self, args=None, kwargs=None, **task_options):
         worker = ray._private.worker.global_worker
         worker.check_connected()
 
+        # We cannot do this when the function is first defined, because we need
+        # ray.init() to have been called when this executes
+        if self._function_signature is None:
+            # Since variable assignment is atomic, we don't need to acquire the
+            # lock before checking _function_signature.
+            self._inject_mut.acquire()
+            self._function = _inject_tracing_into_function(self._function)
+            self._function_signature = ray._private.signature.extract_signature(
+                self._function
+            )
+            self._inject_mut.release()
```

Comment:
should use `with self._inject_mut:` here instead of acquire/release so the lock is released if an exception is raised in either of these calls (also just standard pattern)